synapse/handlers/device.py

```diff
@@ -673,13 +678,135 @@ async def store_dehydrated_device(
             device_id,
             initial_device_display_name,
         )
-        old_device_id = await self.store.store_dehydrated_device(
-            user_id, device_id, device_data
-        )
+
+        time_now = self.clock.time_msec()
+
+        if keys_for_device:
+            keys = await self._check_and_prepare_keys_for_dehydrated_device(
+                user_id, device_id, keys_for_device
+            )
+            old_device_id = await self.store.store_dehydrated_device(
+                user_id, device_id, device_data, time_now, keys
+            )
+        else:
+            old_device_id = await self.store.store_dehydrated_device(
+                user_id, device_id, device_data, time_now
+            )
+
         if old_device_id is not None:
             await self.delete_devices(user_id, [old_device_id])
+
         return device_id
 
+    async def _check_and_prepare_keys_for_dehydrated_device(
+        self, user_id: str, device_id: str, keys: JsonDict
+    ) -> dict:
+        """
+        Check if any of the provided keys are duplicate and raise if they are,
+        prepare keys for insertion in DB
+
+        Args:
+            user_id: user to store keys for
+            device_id: the dehydrated device to store keys for
+            keys: the keys - device_keys, onetime_keys, or fallback keys to store
```

Comment:
This could probably contain a bit more information about the form of this.

Reply:
I've updated it, let me know if it's more useful?